python/Ganga/GPIDev/Credentials2/CredentialStore.py

```diff
@@ -0,0 +1,259 @@
+from __future__ import absolute_import
+
+import collections
+from datetime import timedelta
+
+import Ganga.Utility.logging
+from Ganga.GPIDev.Base.Objects import GangaObject
+from Ganga.GPIDev.Base.Proxy import export
+from Ganga.GPIDev.Schema import Schema, Version
+
+from .exceptions import CredentialsError
+from .ICredentialRequirement import ICredentialRequirement
+from .ICredentialInfo import ICredentialInfo
+
+logger = Ganga.Utility.logging.getLogger()
+
+
+class CredentialStore(GangaObject, collections.Mapping):
+    """
+    The central management for all credentials
+
+    It is not intended to store the credential objects between sessions,
+    rather it will search the filesystem or create new credential files.
+
+    Its interface is similar to a dictionary with ``ICredentialRequirement``
+    objects as keys and ``ICredentialInfo`` objects as values.
+
+    A single instance of this class makes most sense and should be created in the bootstrap and exported.
+    """
+
+    _schema = Schema(Version(1, 0), {})
+
+    _category = 'credentials2'
+    _hidden = 1  # This class is hidden since we want a 'singleton' created in the bootstrap
+
+    def __init__(self):
+        super(CredentialStore, self).__init__()
+        self.credentials = set()
+
+    @export
+    def create(self, query, create=True, check_file=False):
+        # type: (ICredentialRequirement, bool, bool) -> ICredentialInfo
+        """
+        Create an ``ICredentialInfo`` for the query.
+
+        Args:
+            query (ICredentialRequirement):
+            check_file: Raise an exception if the file does not exist
+            create: Create the credential file
+
+        Returns:
+            The newly created ICredentialInfo object
+        """
+
+        cred = query.info_class(query, check_file=check_file, create=create)
+        self.credentials.add(cred)
+        return cred
+
+    def remove(self, credential_object):
+        # type: (ICredentialInfo) -> None
+        """
+        Args:
+            credential_object (ICredentialInfo):
+        """
+
+        self.credentials.remove(credential_object)
+
+    def __str__(self, interactive=False):
+        # Get the string information with which to fill the table
+        headers = ['Type', 'Location', 'Valid', 'Time left']
+        cred_info = [[str(f) for f in (type(cred).__name__, cred.location, cred.is_valid(), cred.time_left())] for cred in self.credentials]
```

Comment:
I can certainly try to format it in a nicer way. The code in this function could perhaps be pulled out into a function in utils since most of it is just doing table formatting?